### PR TITLE
[None][perf] Vectorize compute_logprobs post-processing

### PR DESCRIPTION
**Pull request by zhangxin81**
## Description

compute_logprobs() post-processes logits on the executor response path. Its current implementation performs scalar .item() calls while walking every token position and, for top-k results, every returned entry. On CUDA, each scalar extraction can synchronize the device stream, making prompt logprobs and TRT generation logprobs increasingly expensive as sequence length grows.

This change batches the tensor work and host transfers while preserving the existing output contract:

- Gather sampled-token logprobs for all positions in one operation and transfer them with one .tolist() call.
- Compute sampled-token ranks in batches, chunked by a fixed element budget to bound the temporary [chunk, vocab] comparison tensor.
- Transfer top-k values and token IDs in bulk, and compute the sampled-token value/rank only for rows where it is absent from top-k.
- Preserve simple-format results, dict-format results, the tokens truncation workaround for nvbug 5324291, and the existing rank definition.

## Test Coverage

- Added tests/unittest/executor/test_compute_logprobs.py, covering generation and prompt paths, simple and dict formats, [T, V] and [1, T, V] logits, top-k sampled tokens inside/outside the returned top-k set, missing tokens, and the truncation workaround.
- uvx pre-commit run --files tensorrt_llm/executor/result.py tests/unittest/executor/test_compute_logprobs.py passed, including isort, YAPF, autoflake, ruff, ruff-format, codespell, and legacy lint checks.
- python3 -m py_compile tensorrt_llm/executor/result.py tests/unittest/executor/test_compute_logprobs.py passed.
- CPU differential validation against the original per-token implementation: 54,400 combinations matched exactly.
- GPU pytest is intended for CI; the development host is macOS without CUDA/PyTorch GPU support.

## Performance

The supplied CPU benchmark isolates Python-loop and dispatch overhead and reports the following best-of-three results:

| Tokens | Vocab | top_k | Format | Before | After | Speedup |
| ---: | ---: | ---: | :---: | ---: | ---: | ---: |
| 512 | 32,000 | 0 | simple | 1.98 ms | 0.05 ms | 39.3x |
| 512 | 32,000 | 0 | dict | 27.41 ms | 3.79 ms | 7.2x |
| 512 | 32,000 | 20 | dict | 70.30 ms | 23.10 ms | 3.0x |
| 1,024 | 128,256 | 0 | simple | 3.97 ms | 0.08 ms | 47.9x |
| 1,024 | 128,256 | 0 | dict | 126.62 ms | 29.92 ms | 4.2x |
| 4,096 | 128,256 | 0 | simple | 16.19 ms | 0.32 ms | 50.6x |
| 4,096 | 128,256 | 0 | dict | 624.45 ms | 233.54 ms | 2.7x |

These are CPU-only measurements from the supplied benchmark and do not include CUDA synchronization cost. The primary expected CUDA benefit is removing per-token device-to-host synchronizations; GPU performance should be confirmed by CI or a maintainer GPU run.

## Risk / Follow-up

The rank comparison uses a bounded scratch tensor with a CUDA budget of 1 << 24 boolean elements and a CPU budget of 1 << 20. The CUDA value is intentionally isolated as a constant so it can be tuned after GPU measurement without changing semantics.

## Dev Engineer Review

- `compute_logprobs()` now batches sampled-token gathers and host transfers.
- Rank computation uses bounded chunks and skips tokens already present in top-k results.
- Existing simple and dictionary output formats, rank semantics, token truncation handling, and public APIs remain unchanged.
- The optimization reduces device-to-host synchronization. CUDA performance still needs validation.

## QA Engineer Review

- `tests/unittest/executor/test_compute_logprobs.py` adds CUDA-gated coverage for generation and prompt paths.
- Tests cover batching, tensor shapes, simple and dictionary formats, top-k results, missing sampled tokens, token inclusion, no-token cases, and truncation behavior.
- The tests compare results with a per-token reference implementation.
- No matching CI or manual-QA test-list entry was identified.
- Coverage verdict: `sufficient`, with CUDA performance validation still requiring follow-up.

## Per-File QA Perspective

- `tensorrt_llm/executor/result.py`: Verify batched gather behavior, bounded rank computation, missing-token handling, output format compatibility, and nvbug 5324291 truncation behavior. Confirm CUDA synchronization and memory behavior under representative workloads.
- `tests/unittest/executor/test_compute_logprobs.py`: Covers generation and prompt behavior against a reference implementation across top-k, batching, formats, missing tokens, and truncation. No corresponding `test-db/` or `qa/` list entry was identified.

### CODE REVIEW

**Comment by karljang** [comment]
@tongyuantongyu , could you please help review this PR, thank you!

File: tensorrt_llm/executor/result.py [hunk 1]
```diff
@@ -1252,6 +1252,34 @@ async def __anext__(self):
             raise StopAsyncIteration
 
 
+# Element budget for the chunked rank computation below. The comparison
```
Comment (tongyuantongyu) [inline tensorrt_llm/executor/result.py:1255]: Do you have real data that shows meaningful memory usage reduction and negligible slowdown on this chunking?
We already have the full `[num_tokens, vocab_size]` **float** tensor `logprobs` and `logits` materialized in memory along with the temporaries when computing `log_softmax`, so another `[num_tokens, vocab_size]` boolean tensor doesn't sound like something that must be chunked.

**Review by tongyuantongyu — COMMENTED** [review]
This PR itself is an incremental improvement.

But now with the deprecation of TensorRT backend, these code is only used for prompt logprobs, while PyTorch backend's sampler already has a similar but more advanced implementation dealing with that of generation tokens:

https://github.com/NVIDIA/TensorRT-LLM/blob/main/tensorrt_llm/_torch/pyexecutor/sampler/logprobs.py

This raises the question whether we should consolidate our logprobs handling code. CC @QiJune should we make this a runtime improvement task, or we already have folks working on that?

File: tensorrt_llm/executor/result.py (hunk 1 above)
Comment (zhangxin81) [reply tensorrt_llm/executor/result.py:1255]: Thanks, this is a fair point. The main optimization in this PR is batching the tensor work / host transfers and avoiding per-token `.item()` synchronizations. The chunking is only meant to cap the extra rank-comparison scratch tensor.

I agree the current CPU-only numbers do not prove the CUDA memory/latency tradeoff for the chunked rank path. I will add a focused CUDA benchmark comparing chunked vs unchunked vectorized rank computation on representative prompt-logprobs shapes, and report both peak CUDA memory and latency. If the slowdown is not negligible or the memory saving is not meaningful in practice, I can simplify this part and keep only the bulk-transfer/vectorization changes.

**Comment by zhangxin81** [comment]
On the consolidation point: I agree that consolidating logprobs handling with the PyTorch backend implementation is the right longer-term direction.

For this PR, I tried to keep the scope limited to the existing executor-side `compute_logprobs()` path, which is still used for prompt logprobs and for generation logits when the backend does not already provide sampler-side logprobs. The PyTorch sampler path is more integrated with request/beam layout and staged host copies, so moving this path there would be a broader runtime refactor than this targeted post-processing optimization.

My preference is to keep this PR scoped to the current executor-side improvement, add the focused CUDA data for the chunking tradeoff, and treat the full logprobs consolidation as a follow-up runtime task if maintainers agree.

File: tensorrt_llm/executor/result.py [hunk 2]
```diff
@@ -1252,6 +1252,34 @@ async def __anext__(self):
...
+
+    Equivalent to ``(logprobs[t] > sampled_logprobs[t]).sum() + 1`` per row,
+    but batched over rows and chunked to bound peak memory.
+    """
+    num_tokens, vocab_size = logprobs.shape
+    budget = (_RANK_CHUNK_ELEMS_CUDA
+              if logprobs.device.type == "cuda" else _RANK_CHUNK_ELEMS_CPU)
+    chunk = max(1, budget // max(vocab_size, 1))
+    ranks = torch.empty(num_tokens, dtype=torch.long, device=logprobs.device)
+    for start in range(0, num_tokens, chunk):
```
[the remark this answers is not in the document]
Comment (zhangxin81) [reply tensorrt_llm/executor/result.py:1275]: Agreed, this is a useful regression case. I will add a focused test that monkeypatches the rank chunk budget to force multiple chunks, with sampled tokens outside top-k, and compare the result against the per-token reference implementation.

**Comment by tongyuantongyu** [comment]
@zhangxin81 Yes we are open to have this PR merged first. Please address the review comments, and we'll trigger the CI then.
